# Photography Business Analysis: SQL Insights and Project Overview <a class='anchor' id='top'></a>
## Author: Oliverius, Miranda

## Table of Contents:

* [Introduction](#intro)
* [Project Highlights](#highlights)
* [Preliminaries](#prelims)
* [SQL File Preparation](#prep)
* [Data Analysis Results](#results)
* [Conclusion](#conclusion)

## Introduction <a class='anchor' id='intro'></a>

This project established a relational database for a real estate photography business and leveraged it to gather business intelligence insights. Prior to the database creation, business data was only accessible through reports generated by the client portal, the online platform used for appointment bookings and file deliveries. This system was not designed for business analytics, limiting the owner’s ability to answer performance-related questions or utilize data analytics insights for process improvements. The establishment of a relational database provides a robust foundation for addressing critical business questions and facilitating data-driven decision-making. Project highlights are summarized below.

This notebook connects to the PostgreSQL database and executes business analytics SQL query scripts, with answers noted in the accompanying markdown cells. This structure provides viewers with a clear and organized way to access the business intelligence results of this project. The database creation and SQL script were developed independently, while the implementation in Jupyter Notebook was supported by OpenAI's ChatGPT to enhance accessibility and clarity for viewers.

All data used in this analysis has been anonymized to ensure client confidentiality, with the process for anonymization documented in the GitHub repository linked below.

Additional project files can be accessed at the following links:
* For the complete database creation, import, and cleaning SQL files, visit the <u><a href="https://github.com/databymir/photo_business">GitHub Repository</a></u>.
* The interactive <u><a href="https://public.tableau.com/app/profile/miranda.oliverius/viz/PhotographyBusinessDashboardStory/Dashboard">dashboard</a></u> and <u><a href="https://public.tableau.com/app/profile/miranda.oliverius/viz/PhotographyBusinessDashboardStory/Story">story</a></u> are hosted on Tableau Public.

## Project Highlights <a class='anchor' id='highlights'></a>

[Click Here to Return to Top](#top)

* Designed a relational database with constraints for a real estate photography business using PostgreSQL and pgAdmin.
* Leveraged Python to anonymize Personally Identifiable Information (PII) utilizing the Faker and Pandas packages.
* Imported business data using temporary tables to streamline preliminary data preprocessing.
* Utilized SQL queries for data cleaning and analysis, showcasing various SQL features such as subqueries, joins, aggregation, casting, aliasing, ordering, grouping, and case statements.
* Connected the PostgreSQL database to Tableau.
* Leveraged custom calculated fields, Level of Detail (LOD) calculations, table calculations, and custom parameters to enhance the end-user experience through interactivity.
* Implemented custom parameters and calculated fields that adjusted the granularity of KPI line graphs based on selected relative date ranges (e.g., "Last Year" displays a quarterly line graph, while "Last 90 Days" shows a monthly line graph).
* Created an interactive dashboard to communicate key performance indicators (KPIs) and suggest follow-up opportunities for clients with only one booked photoshoot.
* Crafted a Tableau story to present key insights and introduce the new sales dashboard to the business owner.ness owner

## Preliminaries <a class='anchor' id='prelims'></a>

[Click Here to Return to Top](#top)

To get started, common Python libraries for database connection and data manipulation should be imported. These libraries provide a more flexible and purpose-driven approach to working with data than using base Python alone.

In [1]:
# load libraries
## *** DATABASE CONNECTION ***
import sqlalchemy
from sqlalchemy import create_engine, text
import psycopg2

## *** DATA MANIPULATION ***
import pandas as pd
import re

An engine is created to establish a connection to the PostgreSQL database. This connection allows for the execution of SQL queries and retrieval of results from the photo_business database.

In [2]:
# create engine
engine = create_engine('postgresql+psycopg2://postgres:7BLCOqm2$JojhXo7KFtQ@localhost:5432/photo_business')

## SQL File Preparation <a class='anchor' id='prep'></a>

[Click Here to Return to Top](#top)

The SQL queries are read from a specified file, separated into individual lines, and filtered to identify valid queries. Multi-line queries are combined for execution. This process was developed to streamline querying for analysis while ensuring clarity and accuracy in execution.

In [3]:
# read SQL queries from file
with open(r'C:/Users/miran/OneDrive/data_science/projects/photo_business/4_data_analysis/analysis_queries.sql', 'r') as file:
    sql_script = file.read()

In [4]:
# split script into lines
lines = sql_script.splitlines()

In [5]:
# define SQL keywords
sql_keywords = ['SELECT', 'INSERT', 'UPDATE', 'DELETE', 'CREATE', 'DROP', 'ALTER', 'FROM', 'WHERE', 'JOIN', 'GROUP BY', 'ORDER BY']

In [6]:
# define function to determine if a line is a valid SQL query
def is_valid_line(line):
    stripped = line.strip()
    # Check for comments or empty lines
    if stripped.startswith('/*') or stripped.startswith('--') or stripped.startswith('---') or not stripped:
        return False
    # Allow lines that start with SQL command keywords
    return any(stripped.upper().startswith(keyword) for keyword in sql_keywords)

In [7]:
# combine multi-line SQL queries into single statements
combined_queries = []
current_query = ""

for line in lines:
    stripped_line = line.strip()
    
    if is_valid_line(stripped_line) or (current_query and not stripped_line.startswith('--') and not stripped_line.startswith('/*')):
        # append to current query
        current_query += stripped_line + " "  
        
        # Check if the line ends with a semicolon, indicating the end of the SQL query
        if stripped_line.endswith(';'):
            # save complete query
            combined_queries.append(current_query.strip())
            # reset for next query
            current_query = ""

In [8]:
# store results for each query
query_results = []

with engine.connect() as conn:
    for query in combined_queries:
        try:
            sql_query = conn.execute(text(query))
            result_df = pd.DataFrame(sql_query.fetchall(), columns=sql_query.keys())
            query_results.append((query, result_df))  # Store query and result
            
        except Exception as e:
            print(f"Error executing query: {query}\nError: {e}\n")

## Data Analysis Results <a class='anchor' id='results'></a>

[Click Here to Return to Top](#top)

Each query is executed individually, with its result displayed in a structured format. Following each result, markdown cells provide insights and answers to the questions posed by the queries.

In [9]:
# display first query and results
query_index = 0
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 1:
SELECT  EXTRACT('year' from order_create_date) AS year, COUNT(DISTINCT site_id) AS total_orders FROM orders GROUP BY year;



year  total_orders
0  2023            31
1  2024            78

Answer: We shot 31 properties during 2023, and have shot 78 properties year-to-date (07/31/2024).

In [10]:
# display second query and results
query_index = 1
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 2:
SELECT 	EXTRACT('year' FROM order_create_date) AS year, EXTRACT('month' FROM order_create_date) AS month, COUNT(order_id) AS total_orders FROM orders GROUP BY year, month ORDER BY year, month;



year month  total_orders
0   2023     5             2
1   2023     6             3
2   2023     7             6
3   2023     8             4
4   2023     9             6
5   2023    10             2
6   2023    11             2
7   2023    12             6
8   2024     1             5
9   2024     2             7
10  2024     3             5
11  2024     4            10
12  2024     5            21
13  2024     6            18
14  2024     7            21

Answer: The business owner was still working part-time for his previous employer until mid-way through September 2023, and 
the slow season for real estate tends to be October through March, explaining the low volume of monthly orders from May 2023
through March 2024. April 2024 saw the highest number of orders up to that point (10 orders), but was lower than expected due
to a slow real estate market. Although uncertainty has continued in the market, the next three months averaged 20 orders per
month, doubling the monthly orders from April.

In [11]:
# display third query and results
query_index = 2
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 3:
SELECT ROUND(SUM(amount) / COUNT(DISTINCT order_id), 2) AS avg_order_value FROM invoices;



avg_order_value
0          378.57

Answer: $378.57

In [12]:
# display fourth query and results
query_index = 3
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 4:
SELECT ROUND((CAST(SUM(photo_qty) AS numeric) / COUNT(DISTINCT order_id)), 2) AS avg_photos_per_order FROM invoices;



avg_photos_per_order
0                70.71

Answer: 70.71 photos/order

In [13]:
# display fifth query and results
query_index = 4
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 5:
SELECT  task_category, COUNT(task_category) FROM order_tasks WHERE task_name != 'Dont need a drone?' GROUP BY task_category ORDER BY COUNT(task_category) DESC FETCH NEXT 5 ROWS WITH TIES;



task_category  count
0  Residential Photography     75
1                    Video     27
2                   Custom     21
3                    Drone     20
4         2D/3D Floorplans     19

Answer: Residential Photography, Video, Custom, Drone, and 2D/3D Floorplans 

In [14]:
# display sixth query and results
query_index = 5
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 6:
SELECT  task_name, COUNT(task_name) FROM order_tasks WHERE task_name != 'Dont need a drone?' GROUP BY task_name ORDER BY COUNT(task_name) DESC FETCH NEXT 10 ROWS WITH TIES;



task_name  count
0                     40 Photos     40
1                     30 Photos     25
2                        Custom     21
3                  2D Floorplan     19
4   Basic HD Video Walk-through     16
5   Unlimited Photos-SqFt Based     11
6          Drone Photos & Video     10
7          Social Media Upgrade      8
8          Twilight Photography      7
9      Drone Photos & Exteriors      6
10            8.5x11 Design Fee      6

Answer: 40 Photos, 30 Photos, Custom, 2D Floorplan, Basic HD Video Walk-through, Unlimited Photos-SqFt Based,
Drone Photos & Video, Social Media Upgrade, Twilight Photography, and Drone Photos & Exteriors (tied with 8.5x11 Design Fee)

In [15]:
# display seventh query and results
query_index = 6
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 7:
SELECT  city, COUNT(city) FROM project_sites GROUP BY city ORDER BY COUNT(city) DESC FETCH NEXT 3 ROWS WITH TIES;



city  count
0  Fort Collins     27
1       Greeley     15
2      Loveland     13
3       Windsor     13

Answer: Fort Collins, Greeley, and Loveland (tied with Windsor)

In [16]:
# display eighth query and results
query_index = 7
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 8:
SELECT  city, COUNT(city) FROM project_sites GROUP BY city ORDER BY COUNT(city) FETCH NEXT 3 ROWS WITH TIES;



city  count
0        Evans      1
1  Platteville      1
2     Longmont      1
3      Timnath      1
4     Berthoud      1
5   Estes Park      1
6       Kersey      1

Answer: The least frequent cities are all tied with one photoshoot each- Evans, Platteville, Longmont, Timnath, Berthoud, 
Estes Park, and Kersey

In [17]:
# display ninth query and results
query_index = 8
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 9:
SELECT 	CAST(task_complete_date AS date), COUNT(DISTINCT site_id) AS project_sites_visited FROM order_tasks JOIN orders USING(order_id) JOIN project_sites USING(site_id) GROUP BY CAST(task_complete_date AS date) ORDER BY COUNT(DISTINCT site_id) DESC FETCH NEXT 1 ROWS WITH TIES;



task_complete_date  project_sites_visited
0         2024-06-09                      4

Answer: The maximum number of project sites visited in a single day is four, which happened on 06/09/2024

In [18]:
# display tenth query and results
query_index = 9
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 10:
SELECT 	CAST(task_complete_date AS date), COUNT(DISTINCT city) AS cities_visited FROM order_tasks JOIN orders USING(order_id) JOIN project_sites USING(site_id) GROUP BY CAST(task_complete_date AS date) ORDER BY 	COUNT(DISTINCT city) DESC FETCH NEXT 1 ROWS WITH TIES;



task_complete_date  cities_visited
0         2024-05-28               3

Answer: The maximum number of cities visited in a single day is three, which happened on 05/28/2024.

In [19]:
# display eleventh query and results
query_index = 10
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 11:
SELECT  client_id, client_name, COUNT(order_id) AS total_orders FROM clients JOIN project_sites USING(client_id) JOIN orders USING(site_id) GROUP BY client_id ORDER BY COUNT(order_id) DESC FETCH NEXT 5 ROWS WITH TIES;



client_id        client_name  total_orders
0    432454    Andrew Mitchell            11
1    356353      Jessica Ramos            10
2    444871         David Hill            10
3    370084  Elizabeth Schultz            10
4    333619     David Thompson             9
5    350951    Ashley Cantrell             9

Answer: Andrew Mitchell, Jessica Ramos, David Hill, Elizabeth Schultz, and David Thompson (tied with Ashley Cantrell)

In [20]:
# display twelfth query and results
query_index = 11
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 12:
SELECT  client_id, client_name, SUM(amount) AS total_sales FROM clients JOIN project_sites USING(client_id) JOIN orders USING(site_id) JOIN invoices USING(order_id) GROUP BY client_id ORDER BY total_sales DESC FETCH NEXT 5 ROWS WITH TIES;



client_id        client_name total_sales
0    333619     David Thompson     5225.00
1    370084  Elizabeth Schultz     4965.00
2    350951    Ashley Cantrell     4056.25
3    356353      Jessica Ramos     3660.00
4    432454    Andrew Mitchell     3440.00

Answer: David Thompson, Elizabeth Schultz, Ashley Cantrell, Jessica Ramos, and Andrew Mitchell

In [21]:
# display thirteenth query and results
query_index = 12
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 13:
SELECT 	client_id, client_name, MAX(order_create_date) AS last_order_date FROM clients JOIN project_sites USING(client_id) JOIN orders USING(site_id) GROUP BY client_id HAVING  COUNT(order_id) = 1 AND MAX(order_create_date) > '2023-12-31' ORDER BY MAX(order_create_date) DESC;



client_id       client_name     last_order_date
0    506943    Brooke Pearson 2024-07-11 09:36:00
1    516578     Carlos Vaughn 2024-06-20 13:24:00
2    478730   Michael Aguirre 2024-05-23 17:14:00
3    503349      Debra Austin 2024-05-21 11:24:00
4    478070  Brianna Lawrence 2024-05-14 16:20:00
5    463909     Ashley Willis 2024-04-10 19:20:00

Answer: There are six clients who have only booked with us once but tried our services this year. They are Brooke Pearson,
Carlos Vaughn, Michael Aguirre, Debra Austin, Brianna Lawrence, and Ashley Willis.

In [22]:
# display fourteenth query and results
query_index = 13
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 14:
SELECT ROUND(AVG(payment_date - invoice_date), 2) AS days_outstanding FROM invoices;



days_outstanding
0             9.86

Answer: 9.86 days

In [23]:
# display fifteenth query and results
query_index = 14
query, result_df = query_results[query_index]

print(f"Query {query_index + 1}:\n{query}\n")
display(result_df)

Query 15:
SELECT 	client_id, client_name, ROUND(AVG(payment_date - invoice_date), 2) AS days_outstanding FROM invoices JOIN orders USING(order_id) JOIN project_sites USING(site_id) JOIN clients USING(client_id) GROUP BY client_id, client_name ORDER BY days_outstanding DESC;



client_id        client_name days_outstanding
0     378999       Casey Morris            37.33
1     407799      Stacy Delgado            24.00
2     349192         Lori Perez            20.33
3     367670    Carly Rodriguez            16.00
4     506943     Brooke Pearson            15.00
5     475543         Mark Davis            12.00
6     444871         David Hill            11.30
7     432454    Andrew Mitchell            11.29
8     407685           Lisa Lee            11.00
9     444328      Tracy Andrade            11.00
10    409249     Lawrence Ortiz            10.50
11    333619     David Thompson             8.43
12    421913       Brian Torres             8.40
13    350951    Ashley Cantrell             8.22
14    373658        Adam Pierce             8.00
15    503349       Debra Austin             7.00
16    516578      Carlos Vaughn             7.00
17    370084  Elizabeth Schultz             6.30
18    463909      Ashley Willis             6.00
19    478070   Brianna Lawrence             6.00
20    504038    Caroline Hodges             6.00
21    356353      Jessica Ramos             5.60
22    507883    Brittney Sawyer             5.50
23    425974    Nicholas Spence             5.50
24    478730    Michael Aguirre             5.00
25    334554       Amanda Poole             5.00
26    355103   Heather Martinez             4.80
27    520182     Wayne Marshall             4.33
28    376264         Tammy Cook             4.33
29    341721        Erin Arnold             0.00

Answer: Casey Morris takes the longest to pay, averaging 37.33 days between invoice and payment date.
Meanwhile, Erin Arnold is the quickest to pay, averaging 0.00 days between invoice and payment date.

## Conclusion <a class='anchor' id='conclusion'></a>

[Click Here to Return to Top](#top)

This analysis has yielded valuable insights into the performance of the real estate photography business. The executed queries have uncovered key metrics, including average order values, payment timelines, and popular service offerings. These findings highlight areas for potential improvement and support data-driven decision-making. By leveraging the established relational database, the business can effectively track its performance and implement strategies for growth and efficiency. Following this SQL analysis, the database was connected to Tableau Desktop, resulting in the interactive dashboard and story linked in the introduction.